internal/juju/jaas.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the Apache License, Version 2.0, see LICENCE file for details.
+
+package juju
+
+import (
+	"github.com/canonical/jimm-go-sdk/v3/api"
+	"github.com/canonical/jimm-go-sdk/v3/api/params"
+	jujuapi "github.com/juju/juju/api"
+
+	"github.com/juju/errors"
+)
+
+type jaasClient struct {
+	SharedClient
+	getJaasApiClient func(jujuapi.Connection) JaasAPIClient
+}
+
+func newJaasClient(sc SharedClient) *jaasClient {
+	return &jaasClient{
+		SharedClient: sc,
+		getJaasApiClient: func(conn jujuapi.Connection) JaasAPIClient {
+			return api.NewClient(conn)
+		},
+	}
+}
+
+// AddRelations attempts to create the provided slice of relationship tuples.
+func (jc *jaasClient) AddRelations(tuples []params.RelationshipTuple) error {
+	conn, err := jc.GetConnection(nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+	cl := jc.getJaasApiClient(conn)
+	req := params.AddRelationRequest{
+		Tuples: tuples,
+	}
+	return cl.AddRelation(&req)
+}
+
+// DeleteRelations attempts to delete the provided slice of relationship tuples.
+func (jc *jaasClient) DeleteRelations(tuples []params.RelationshipTuple) error {
+	conn, err := jc.GetConnection(nil)
+	if err != nil {
+		return err
+	}
+	defer func() { _ = conn.Close() }()
+	cl := jc.getJaasApiClient(conn)
+	req := params.RemoveRelationRequest{
+		Tuples: tuples,
+	}
+	return cl.RemoveRelation(&req)
+}
+
+// ReadRelations attempts to read relations that match the criteria defined by `tuple`.
+func (jc *jaasClient) ReadRelations(tuple *params.RelationshipTuple) ([]params.RelationshipTuple, error) {
+	if tuple == nil {
+		return nil, errors.New("add relation request nil")
+	}
+
+	conn, err := jc.GetConnection(nil)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = conn.Close() }()
+
+	client := jc.getJaasApiClient(conn)
+	relations := make([]params.RelationshipTuple, 0)
+	req := &params.ListRelationshipTuplesRequest{Tuple: *tuple}
+	for {
```

Comment:
issue: Use of an infinite loop. Use the juju/retry package or something else. This ignores termination from terraform via their context.

Reply:
Instead of an arbitrary number, you can pass the context in from the provider code to this method. The context is provided by terraform. Check context.Done() at the end of the for loop. This will allow terraform to exert control if an early exit is needed. 

```
	select {
	case <-ctx.Done():
	        return nil, ctx.Err()
	default:
	}
```

Reply:
Good idea, done.